PROTOCOL.md

```diff
@@ -0,0 +1,193 @@
+# Twirp Wire Protocol
+
+This document defines the Twirp wire protocol over HTTP.
+
+## Conventions
+
+The requirement level keywords "MUST", "MUST NOT", "REQUIRED",
+"SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY",
+and "OPTIONAL" used in this document are to be interpreted as
+described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+The grammar rules used in this document are using [ABNF
+syntax](https://tools.ietf.org/html/rfc5234).
+
+## Overview
+
+The Twirp wire protocol is a simple RPC protocol based on HTTP and
+Protocol Buffers (proto). The protocol uses HTTP URLs to specify the
+RPC endpoints, and sends/receives proto messages as HTTP
+request/response bodies.
+
+To use Twirp, developers first define their APIs using proto files,
+then use Twirp tools to generate the client and the server libraries.
+The generated libraries implement the Twirp wire protocol, using the
+standard HTTP library provided by the programming language runtime or
+the operating system. Once the client and the server are implemented,
+the client can communicate with the server by making RPC calls.
+
+The Twirp wire protocol supports both binary and JSON encodings of
+proto messages, and works with any HTTP client and any HTTP version.
+However, certain capabilities may be limited by the actual HTTP
+library being used.
+
+### URLs
+
+**URL ::= Base-URL "/" [ Package "." ] Interface "/" Method**
+
+The Twirp wire protocol uses HTTP URLs to directly specify the RPC
+endpoints on the server for sending the requests. such direct mapping
+makes the request routing simple and efficient. The Twirp URLs have
+the following components.
+
+* **Base-URL** is the virtual location of a Twirp API server, which is
+  typically published via API documentation or service discovery.  For
+  example, "https://example.com/apis".
```

Comment:
This is a bad example right now, as the base URL in v5 cannot safely include a path component.

I think we should be more specific here. The Base URL should contain only the "scheme" and "authority" (commonly known as "host and port") portions of a [RFC 3986 URI](https://tools.ietf.org/html/rfc3986).

We will likely change this to allow a path component in v6, as discussed in #55.